### PR TITLE
Add text field support to archive indices

### PR DESCRIPTION
**Pull request by ywelsch**
Adds support for "text" fields in archive indices, with the goal of adding simple filtering support on text fields when querying archive indices.

There are some differences to regular text fields:
- no global statistics: queries on text fields return constant score (similar to `match_only_text`).
- `analyzer` fields can be updated
- if defined `analyzer` is not available, falls back to default `analyzer`
- no guarantees that analyzers are BWC

The above limitations also give us the flexibility to eventually swap out the implementation with a "runtime-text field" variant, and hence only provide those capabilities that can be emulated via a runtime field.

Relates #81210

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java [hunk 1]
```diff
@@ -92,7 +89,7 @@ public IndexMetadata verifyIndexMetadata(IndexMetadata indexMetadata, Version mi
         // Next we have to run this otherwise if we try to create IndexSettings
         // with broken settings it would fail in checkMappingsCompatibility
         newMetadata = archiveBrokenIndexSettings(newMetadata);
-        createAndValidateMapping(newMetadata);
+        checkMappingsCompatibility(newMetadata);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java:92]: The changes in this class revert the changes made in https://github.com/elastic/elasticsearch/pull/85059, as we now validate against a MapperService with all analyzers configured when restoring the legacy index in `RestoreService`. The reason for doing it this way now is that it provides better error messages on restore, but also handles a tricky situation where the Mapping returned by these methods here would have their analyzer settings misconfigured as checkMappingsCompatibility would not create a proper environment with actual analyzers configured.

File: server/src/test/java/org/elasticsearch/index/mapper/ConstantScoreTextFieldTypeTests.java [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.apache.lucene.search.MultiTermQuery.CONSTANT_SCORE_REWRITE;
+import static org.hamcrest.Matchers.equalTo;
+
+public class ConstantScoreTextFieldTypeTests extends FieldTypeTestCase {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/index/mapper/ConstantScoreTextFieldTypeTests.java:45]: This contains all the tests of `TextFieldTypeTests` with some adaptations for constant scoring.

File: server/src/test/java/org/elasticsearch/index/mapper/MultiFieldTests.java [hunk 3]
```diff
@@ -227,7 +227,7 @@ public void testUnmappedLegacyFieldsUnderKnownRootField() throws Exception {
             b.startObject("name");
             b.field("type", "keyword");
             b.startObject("fields");
-            b.startObject("subfield").field("type", "text").endObject();
+            b.startObject("subfield").field("type", CompletionFieldMapper.CONTENT_TYPE).endObject();
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/index/mapper/MultiFieldTests.java:230]: now that we support text fields, we can't use it anymore to check "placeholder" functionality. Instead we use another unsupported field.

**Review by romseygeek — COMMENTED** [review]
Thanks @ywelsch! This looks pretty close - I left a couple of questions.

File: server/src/test/java/org/elasticsearch/index/mapper/ConstantScoreTextFieldTypeTests.java [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    private static ConstantScoreTextFieldType createFieldType() {
+        return new ConstantScoreTextFieldType("field");
+    }
+
+    public void testIsAggregatableDependsOnFieldData() {
+        ConstantScoreTextFieldType ft = createFieldType();
+        assertFalse(ft.isAggregatable());
+        ft.setFielddata(true);
+        assertTrue(ft.isAggregatable());
+    }
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/index/mapper/ConstantScoreTextFieldTypeTests.java]: Disabling fielddata on older indexes is ok I think (the main purpose, as stated in the PR description, is to provide basic filtering support on archive indices). I've addressed this in [56d391c](https://github.com/elastic/elasticsearch/pull/86591/commits/56d391c364c486ee935aaad75d351fe241d0b040)

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 5]
```diff
@@ -1042,23 +1047,48 @@ public static <T extends Enum<T>> Parameter<T> restrictedEnumParam(
...
-            Supplier<NamedAnalyzer> defaultAnalyzer
+            Supplier<NamedAnalyzer> defaultAnalyzer,
+            Version indexCreatedVersion
         ) {
             return new Parameter<>(name, updateable, defaultAnalyzer, (n, c, o) -> {
                 String analyzerName = o.toString();
                 NamedAnalyzer a = c.getIndexAnalyzers().get(analyzerName);
                 if (a == null) {
-                    throw new IllegalArgumentException("analyzer [" + analyzerName + "] has not been configured in mappings");
+                    if (indexCreatedVersion.isLegacyIndexVersion()) {
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:1063]: I had looked into this (and gave up). One of the challenges was that the `defaultAnalyzer` is defined here when instantiating the `analyzerParam` (i.e. per field), not in the `IndexAnalyzers` (i.e. per index). This means that returning the field-specific default analyzer is not possible.

File: server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java [hunk 6]
```diff
@@ -897,6 +909,78 @@ public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, S
...
+            return new ConstantScoreQuery(super.multiPhraseQuery(stream, slop, enablePositionIncrements, queryShardContext));
+        }
+
+        @Override
+        public Query phrasePrefixQuery(TokenStream stream, int slop, int maxExpansions, SearchExecutionContext queryShardContext)
+            throws IOException {
+            // Disable scoring
+            return new ConstantScoreQuery(super.phrasePrefixQuery(stream, slop, maxExpansions, queryShardContext));
+        }
+
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java:983]: I agree that we don't need to support them. I've pushed [604d70c](https://github.com/elastic/elasticsearch/pull/86591/commits/604d70c05b677de421b18d6ab4849efa24b02607) but I'm not sure how to add a test for it (I couldn't find existing tests that exercise this method).

**Review by romseygeek — APPROVED** [review]
LGTM, thanks @ywelsch

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

**Review by javanna — APPROVED** [review]
LGTM

**Comment by ywelsch** [comment]
Thanks @romseygeek and @javanna!

**Pull request merged**